### PR TITLE
osd: reuse coll and object encode resut in get_encoded_bytes

### PR DESCRIPTION
**Pull request by XinzeChi**
In FileStore::queue_transactions, osd should get_num_bytes and
encode tnx. The get_num_bytes function would encode ghobject and
map which is CPU-intensive.

We should ensure that we donot modify the tnx after get_encoded_bytes.
If modify it, we should reset the coll_object_tbl.

There are two way reset the coll_object_tbl. one is calling
reset_coll_object_tbl manually. The other is encoding transactions.
Signed-off-by: Xinze Chi <xinze@xsky.com>

### CODE REVIEW

**Comment by liewegas** [comment]
lgtm!

**Comment by liewegas** [comment]
2015-11-28T14:31:04.642 INFO:teuthology.orchestra.run.burnupi14.stderr:./os/ObjectStore.h: In function 'ObjectStore::Transaction::Op* ObjectStore::Transaction::_get_next_op()' thread 7fca455924c0 time 2015-11-28 14:31:04.656494
2015-11-28T14:31:04.642 INFO:teuthology.orchestra.run.burnupi14.stderr:./os/ObjectStore.h: 893: FAILED assert(!coll_object_tbl.length())
2015-11-28T14:31:04.642 INFO:teuthology.orchestra.run.burnupi14.stderr: ceph version 10.0.0-715-g995637e (995637e717ad35a3c62bf57734c6dd519ef1da69)
2015-11-28T14:31:04.643 INFO:teuthology.orchestra.run.burnupi14.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2015-11-28T14:31:04.643 INFO:teuthology.orchestra.run.burnupi14.stderr:2015-11-28 14:31:04.656526 7fca455924c0 -1 ./os/ObjectStore.h: In function 'ObjectStore::Transaction::Op* ObjectStore::Transaction::_get_next_op()' thread 7fca455924c0 time 2015-11-28 14:31:04.656494
2015-11-28T14:31:04.643 INFO:teuthology.orchestra.run.burnupi14.stderr:./os/ObjectStore.h: 893: FAILED assert(!coll_object_tbl.length())

easy to reproduce with ceph_test_objectstore

**Comment by XinzeChi** [comment]
@liewegas , Because there is not encoding transaction process in keyvaluestore and filestore. So we should call reset_coll_object_tbl manually. This pass ceph_test_objectstore.

File: src/os/newstore/NewStore.cc [hunk 1]
```diff
@@ -2783,6 +2783,8 @@ int NewStore::queue_transactions(
     _txc_add_transaction(txc, *p);
   }
 
+  reset_coll_object_tbl(tls);
```
Comment (liewegas) [inline src/os/newstore/NewStore.cc]: why is this needed?

Comment (XinzeChi) [inline src/os/newstore/NewStore.cc]: Because we must ensure that the transaction is not modified after we use coll_object_tbl to cache the encode result.
But https://github.com/ceph/ceph/blob/master/src/test/objectstore/store_test.cc#L1973https://github.com/ceph/ceph/blob/master/src/test/objectstore/store_test.cc#L1973,  the transaction is added new op (line 1981) after apply_transaction (line 1975) .
So I think when queue_transaction finish, we reset_coll_object_tbl is better.

Actually, coll_object_tbl is only useful for filestore, because newstore and keyvaluestore there is not encode transaction process. But in objectstore, we do not know which type of store backend.

**Comment by XinzeChi** [comment]
https://github.com/ceph/ceph/pull/7775 looks a better way.
